File: UnityMcpServer/src/tools/create_scriptable_object.py
"""
Defines the create_scriptable_object tool for creating ScriptableObjects in Unity.
"""
import asyncio
from typing import Dict, Any
from mcp.server.fastmcp import FastMCP, Context
from unity_connection import get_unity_connection

def register_create_scriptable_object_tools(mcp: FastMCP):
    """Registers the create_scriptable_object tool with the MCP server."""

    @mcp.tool()
    async def create_scriptable_object(
        ctx: Context,
        action: str,
        name: str = None,
        subfolder: str = None,
        type_name: str = None,
        path: str = None
    ) -> Dict[str, Any]:
        """Creates ScriptableObjects in Unity project, specifically designed for AICharacter and AIInteraction objects.

        Args:
            ctx: The MCP context.
            action: The action to perform (create_ai_character, create_ai_interaction, create_custom, list_types).
            name: Name for the ScriptableObject (required for create actions).
            subfolder: Optional subfolder within the type's base directory (e.g., 'AD', 'CL', 'DITO').
            type_name: Custom ScriptableObject type name (required for create_custom action).
            path: Full path for custom ScriptableObject creation (required for create_custom action).

        Returns:
            A dictionary with operation results ('success', 'message', 'data').
        """
        # Prepare parameters for the C# handler
        params_dict = {
            "action": action.lower(),
        }
        
        # Add optional parameters if provided
        if name is not None:
            params_dict["name"] = name
        if subfolder is not None:
            params_dict["subfolder"] = subfolder
        if type_name is not None:
            params_dict["type_name"] = type_name
        if path is not None:
            params_dict["path"] = path

        # Get Unity connection and execute the command
        connection = get_unity_connection()
        result = await asyncio.to_thread(
            connection.send_command, 
            "HandleCreateScriptableObject", 
            params_dict
        )
        
        return result

File: UnityMcpBridge/Editor/Tools/CreateScriptableObject.cs
using System;
using System.IO;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using UnityMcpBridge.Editor.Helpers; // For Response class

namespace UnityMcpBridge.Editor.Tools
{
    /// <summary>
    /// Handles ScriptableObject creation operations within the Unity project.
    /// </summary>
    public static class CreateScriptableObject
    {
        /// <summary>
        /// Main handler for ScriptableObject creation commands
        /// </summary>
        public static object HandleCommand(JObject @params)
        {
            string action = @params["action"]?.ToString()?.ToLower();
            if (string.IsNullOrEmpty(action))
            {
                return Response.Error("Action parameter is required.");
            }

            try
            {
                switch (action)
                {
                    case "create_ai_character":
                        return CreateAICharacter(@params);
                    
                    case "create_ai_interaction":
                        return CreateAIInteraction(@params);
                    
                    case "create_custom":
                        return CreateCustomScriptableObject(@params);
                    
                    case "list_types":
                        return ListAvailableTypes();
                    
                    default:
                        return Response.Error($"Unknown action: {action}. Available actions: create_ai_character, create_ai_interaction, create_custom, list_types");
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"[CreateScriptableObject] Action '{action}' failed: {e}");
                return Response.Error($"Error handling command: {e.Message}");
            }
        }

        private static object CreateAICharacter(JObject @params)
        {
            string name = @params["name"]?.ToString();
            string subfolder = @params["subfolder"]?.ToString() ?? "";
            
            if (string.IsNullOrEmpty(name))
            {
                return Response.Error("Parameter 'name' is required for creating AI Character");
            }

            var result = ScriptableObjectUtility.CreateAICharacter(name, subfolder);
            
            if (result.success)
            {
                return Response.Success(result.message, new
                {
                    assetPath = result.assetPath,
                    type = "AICharacter",
                    name = name,
                    subfolder = subfolder
                });
            }
            else
            {
                return Response.Error(result.message);
            }
        }

        private static object CreateAIInteraction(JObject @params)
        {
            string name = @params["name"]?.ToString();
            string subfolder = @params["subfolder"]?.ToString() ?? "";
            
            if (string.IsNullOrEmpty(name))
            {
                return Response.Error("Parameter 'name' is required for creating AI Interaction");
            }

            var result = ScriptableObjectUtility.CreateAIInteraction(name, subfolder);
            
            if (result.success)
            {
                return Response.Success(result.message, new
                {
                    assetPath = result.assetPath,
                    type = "AIInteraction",
                    name = name,
                    subfolder = subfolder
                });
            }
            else
            {
                return Response.Error(result.message);
            }
        }

        private static object CreateCustomScriptableObject(JObject @params)
        {
            string typeName = @params["type_name"]?.ToString();
            string name = @params["name"]?.ToString();
            string path = @params["path"]?.ToString();
            
            if (string.IsNullOrEmpty(typeName))
            {
                return Response.Error("Parameter 'type_name' is required for creating custom ScriptableObject");
            }
            
            if (string.IsNullOrEmpty(name))
            {
                return Response.Error("Parameter 'name' is required for creating custom ScriptableObject");
            }
            
            if (string.IsNullOrEmpty(path))
            {
                return Response.Error("Parameter 'path' is required for creating custom ScriptableObject");
            }

            var result = ScriptableObjectUtility.CreateScriptableObject(typeName, path, name);
            
            if (result.success)
            {
                return Response.Success(result.message, new
                {
                    assetPath = result.assetPath,
                    type = typeName,
                    name = name,
                    path = path
                });
            }
            else
            {
                return Response.Error(result.message);
            }
        }

        private static object ListAvailableTypes()
        {
            try
            {
                string[] types = ScriptableObjectUtility.GetAvailableScriptableObjectTypes();
                
                return Response.Success($"Found {types.Length} ScriptableObject types", new
                {
                    types = types
                });
            }
            catch (Exception e)
            {
                return Response.Error($"Error listing types: {e.Message}");
            }
        }
    }
}